spec/unit/middleware/timeout_spec.rb

```diff
@@ -0,0 +1,177 @@
+        end
+
+        it 'invokes job when positive timeout specified' do
+          worker = make_worker(InactiveTimeout, 120, kernel_class)
+          worker = worker
+
+          worker.around_perform job
+
+          expect(worker.around_perform_call_counter).to eq(1)
+        end
+
+        context 'when no timeout event detected' do
+          it 'does not invoke timeout recovery' do
+            worker = make_worker(InactiveTimeout, 120, kernel_class)
+
+            expect(kernel_class).not_to receive(:exit!)
+            expect(job).not_to receive(:reconnect_to_redis)
+            expect(job).not_to receive(:fail)
+
+            worker.around_perform job
+          end
+        end
+
+        context 'when timeout event detected' do
+          it 'rescues ::Timeout::Error in case of timeout' do
+            worker = make_worker(TriggeredTimeout, 120, kernel_class)
+            expect {
+              worker.around_perform job
+            }.not_to raise_error
+          end
+
+          it 'reconnects to redis (to recover in case redis causes timeout)' do
+            worker = make_worker(TriggeredTimeout, 120, kernel_class)
+            expect(job).to receive(:reconnect_to_redis)
+            expect {
+              worker.around_perform job
+            }.not_to raise_error
+          end
+
+          it 'fails the job' do
+            worker = make_worker(TriggeredTimeout, 120, kernel_class)
+            expect(job).to receive(:fail)
+            expect {
+              worker.around_perform job
+            }.not_to raise_error
+          end
+
+          context 'when worker does not install RequeueExceptions middleware' do
+            it 'does not call neither #requeueable? nor #handle_exception' do
+              worker = make_worker(TriggeredTimeout, 120, kernel_class)
+
+              expect {
+                worker.around_perform job
+              }.not_to raise_error
+            end
+          end
+
+          context 'when worker installs RequeueExceptions middleware' do
+            it 'requeues job if JobTimedoutError requeue requested' do
+              worker = make_worker(TriggeredTimeout, 120, kernel_class)
+              worker.extend(RequeueExceptions)
+              worker.requeue_on JobTimedoutError, delay_range: (1..2), max_attempts: 3
+              allow(worker).to receive(:handle_exception).with(job, anything)
+
+              expect {
+                worker.around_perform job
+              }.not_to raise_error
+            end
+
+            it 'does not requeue job if JobTimedoutError not configured' do
+              worker = make_worker(TriggeredTimeout, 120, kernel_class)
+              worker.extend(RequeueExceptions)
+              expect(worker).not_to receive(:handle_exception).with(job, anything)
+
+              expect {
+                worker.around_perform job
+              }.not_to raise_error
+            end
+          end
+
+          it 'terminates process' do
```

Comment:
I think this doc string could benefit from some additional explanation:

``` ruby
it 'terminates the process so it is not left in an inconsistent state (since `Timeout` can do that)'
```